lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -73,7 +73,7 @@
                         .<Token<K>, V>removalListener(removal -> {
                             Token<K> token = removal.getKey();
                             verify(token != null, "token is null");
-                            if (removal.getCause() != RemovalCause.REPLACED) {
+                            if (removal.getCause() != RemovalCause.REPLACED && removal.getCause() != RemovalCause.EXPIRED) {
```

Comment:
I drafted a test ensuring this behavior is correct (of course, the test code is not "PR ready" but for a comment should be fine)

```java
    @Test
    public void testNoMemoryLeakWithTimeExpiration()
            throws InterruptedException
    {
        LoadingCache<String, String> cache = EvictableCacheBuilder.newBuilder()
                .expireAfterWrite(1, TimeUnit.MILLISECONDS)
                .build(CacheLoader.from(k -> k + " ala ma kota"));

        for (long i = 0; i < 10_000_000_000L; i++) {
            if (i % 1000 == 0) {
                Thread.sleep(1);
            }
            assertEquals(cache.getUnchecked(Long.toString(i)), i + " ala ma kota");
            assertThat(((EvictableCache<?, ?>) cache).tokensCount()).as("tokensCount")
                    // We insert entries 1k/ms, so with 1ms expiry there should be about at most 1k entries
                    .isLessThan(3000);
        }
    }
```

let me know whether you'd agree with these assertions would be reasonable?

Reply:
Actually, as I'm looking more closely, it doesn't fix it correctly.
Elements can be evicted from the cache if their time expires during read or write operation on a cache segment. So if keys 'foo' and 'bar' end up in the same segment (similar to hashtable buckets), then eviction would happen on attempt to `get()` any of those. That means 'EXPIRE' notification is not equivalent to `REPLACE`.
In my test I've only observed one key which is not enought.